### PR TITLE
fix(forward): prevent busy loop on connection err

### PR DESCRIPTION
**Pull request by thevilledev**
### 1. Why is this pull request needed and what does it do?

Fixes a bug in the forward plugin where an immediate connection failure (e.g., TCP RST) could trigger an infinite busy loop. The retry logic failed to increment the "fails" counter when a connection error occurred, causing the loop condition to remain permanently true. This patch fixes it and adds a regression test.

If you run the test against current `master` you get this:

```
=== RUN   TestForward_Regression_NoBusyLoop
    forward_test.go:135: Expected 1 span, got 6071
--- FAIL: TestForward_Regression_NoBusyLoop (0.50s)
FAIL
```

This confirms that a single DNS query caused 6071 spans, causing syscall thrashing.

### 2. Which issues (if any) are related?

I couldn't find any specific ones. Could be a factor in "High CPU usage" issues like #6090.

### 3. Which documentation changes (if any) need to be made?

None.

### 4. Does this introduce a backward incompatible change or deprecation?

No.

### CODE REVIEW

**Review by yongtang — APPROVED** [review]

File: plugin/forward/forward.go [hunk 1]
```diff
@@ -191,6 +191,7 @@ func (f *Forward) ServeDNS(ctx context.Context, w dns.ResponseWriter, r *dns.Msg
 				proxy.Healthcheck()
 			}
 
+			fails++
```
Comment (rdrozhdzh) [inline plugin/forward/forward.go:194]: I don't like this fix - looks too risky. This will probably break the health check behavior and may break the `max_fails` setting. Not sure about other side effects.
While the problem of "quick error" does exist, this fix doesn't really check if this is a "quick error".

Comment (thevilledev) [reply plugin/forward/forward.go:194]: I see where you're coming from. To clarify, I did test a number of edge cases (including different `max_fails` values and fast vs slow upstream failures). Funnily enough, that exploration is what originally led me to work on this.

Thinking of alternatives. Instead of coupling this to health state, how about we add a per‑request attempt counter inside `ServeDNS`? This would cap the total number of `proxy.Connect` calls for a single request to a set limit, which could be configurable.

Comment (rdrozhdzh) [reply plugin/forward/forward.go:194]: Well, I'm not saying the new behavior is worse, I'm saying it's different. Don't know how critical keeping the old behavior may be for users. 
IIUC, the main difference is that earlier the plugin might loop over the list of proxies multiple times until all of them are considered down (by health checker) or until deadline reached. Not sure if keeping this functionality is important. Probably, this should be discussed with authors and community.

> Thinking of alternatives. Instead of coupling this to health state, how about we add a per‑request attempt counter inside ServeDNS? This would cap the total number of proxy.Connect calls for a single request to a set limit, which could be configurable.

Maybe. But the reasonable value may depend on the number of proxies.
Some time ago when I faced similar issue, I thought of 1) adding a delay when resetting back to the first proxy; 2) capping the number of loops over the proxy list.

The new behavior should be configurable and backward compatible.

Comment (thevilledev) [reply plugin/forward/forward.go:194]: Good points. I don't see it as important but lets have that discussion in a Github issue and hear people out.

What I've done now:

- New `max_connect_attempts` knob in https://github.com/coredns/coredns/pull/7722
- Github issue about changing the default behaviour in https://github.com/coredns/coredns/issues/7723

Thanks for the feedback!
